node/lib/util/stitch_util.js

```diff
@@ -0,0 +1,875 @@
+ * @param {Object}                  urls    submodule name to url
+ * @param {(String) => Boolean}     keepAsSubmodule
+ * @param {(String) => String|null} adjustPath
+ * @pram {TreeUtil.Change}
+ */
+exports.computeModulesFile = co.wrap(function *(repo,
+                                                urls,
+                                                keepAsSubmodule,
+                                                adjustPath) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.isObject(urls);
+    assert.isFunction(keepAsSubmodule);
+    assert.isFunction(adjustPath);
+
+    const keptUrls = {};
+    for (let name in urls) {
+        const adjusted = adjustPath(name);
+        if (null !== adjusted && keepAsSubmodule(name)) {
+            keptUrls[adjusted] = urls[name];
+        }
+    }
+    const modulesText = SubmoduleConfigUtil.writeConfigText(keptUrls);
+    const db = yield repo.odb();
+    const BLOB = 3;
+    const id = yield db.write(modulesText, modulesText.length, BLOB);
+    return new TreeUtil.Change(id, FILEMODE.BLOB);
+});
+
+/**
+ * Return a map from submodule name to shas to list of objects containing the
+ * fields:
+ * - `metaSha` -- the meta-repo sha from which this subodule sha came
+ * - `url`     -- url configured for the submodule
+ * - `sha`     -- sha to fetch for the submodule
+ * this map contains entries for all shas introduced in the specified `toFetch`
+ * list in the specified `repo`.  Note that the behavior is undefined unless
+ * `toFetch` is ordered from least to most dependent commits.  Perform at most
+ * the specified `numParallel` operations in parallel.  Do not process entries
+ * for submodules for which the specified `keepAsSubmodule` returns true or the
+ * specified `adjustPath` returns null.
+ *
+ * @param {NodeGit.Repository}      repo
+ * @param {[NodeGit.Commit]}        toFetch
+ * @param {(String) => Boolean}     keepAsSubmodule
+ * @param {(String) => String|null} adjustPath
+ * @param {Number}                  numParallel
+ * @return {Object}  map from submodule name -> { metaSha, url, sha }
+ */
+exports.listFetches = co.wrap(function *(repo,
+                                         toFetch,
+                                         keepAsSubmodule,
+                                         adjustPath,
+                                         numParallel) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.isArray(toFetch);
+    assert.isFunction(keepAsSubmodule);
+    assert.isFunction(adjustPath);
+    assert.isNumber(numParallel);
+
+    let urls = {};
+
+    // So that we don't have to continuously re-read the `.gitmodules` file, we
+    // will assume that submodule URLs never change.
+
+    const getUrl = co.wrap(function *(commit, sub) {
+        let subUrl = urls[sub];
+
+        // If we don't have the url for this submodule, load them.
+
+        if (undefined === subUrl) {
+            console.log("loading urls");
```

Comment:
Do we mean to keep this output?

Reply:
No, I don't think this is useful output.